packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1524,6 +1535,46 @@ function rerenderTransition(
   return [start, isPending];
 }
 
+function mountOpaqueIdentifier(): OpaqueIDType | void {
+  let hasSetId = false;
+
+  const [id, setId] = mountState(() => {
+    if (getIsHydrating()) {
+      return makeOpaqueHydratingObject(() => {
+        if (!hasSetId) {
+          setId(() => makeClientId());
+          hasSetId = true;
+        }
+      }, getComponentName(currentlyRenderingFiber.type));
+    } else {
+      return makeClientId();
+    }
+  });
+
+  if (currentlyRenderingFiber.mode === NoMode) {
```

Comment:
I don't understand this conditional.

Reply:
AFAIK NoMode refers to Legacy mode. So if it's legacy mode, we don't want the opaque object to throw, so we use an effect to set the ID instead.

Reply:
This assumes we won’t add other modes that are additive to Legacy. In fact I think we already have one — StrictMode. That’s not NoMode but it *is* legacy when on its own. 

I think what you want to check is `(mode & BlockingMode) == NoMode`. That says “doesn’t have BlockingMode bit”. That bit is present in all non-legacy modes so it’s absence means it’s legacy.